Source/AFResult.swift

```diff
@@ -0,0 +1,125 @@
+//
+//  AFResult.swift
+//
+//  Copyright (c) 2014-2019 Alamofire Software Foundation (http://alamofire.org/)
+//
+//  Permission is hereby granted, free of charge, to any person obtaining a copy
+//  of this software and associated documentation files (the "Software"), to deal
+//  in the Software without restriction, including without limitation the rights
+//  to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+//  copies of the Software, and to permit persons to whom the Software is
+//  furnished to do so, subject to the following conditions:
+//
+//  The above copyright notice and this permission notice shall be included in
+//  all copies or substantial portions of the Software.
+//
+//  THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+//  IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+//  FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+//  AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+//  LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+//  OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+//  THE SOFTWARE.
+//
+
+import Foundation
+
+public typealias AFResult<T> = Result<T, Error>
+
+
+// MARK: AFExtension
+
+extension Result {
+    /// Type that acts as a extension point for all `Result` types.
+    public struct AFExtension {
+        let result: Result<Success, Failure>
+    }
+
+    public static var af: AFExtension.Type { return AFExtension.self }
+    public var af: AFExtension { return AFExtension(result: self) }
+}
+
+// MARK: - Public APIs
+
+extension Result.AFExtension {
+    /// Returns the associated value if the result is a success, `nil` otherwise.
+    public var value: Success? {
+        guard case .success(let value) = result else { return nil }
+        return value
+    }
+
+    /// Returns the associated error value if the result is a failure, `nil` otherwise.
+    public var error: Error? {
```

Comment:
Can the return type be the generic type `Failure` instead of `Error`? It would return the specific type that implements `Error` in cases where the `Failure` generic type is more specific than `Error`

Reply:
But that won't be the case with Alamofire as is would it since the `AFResult` is locked to `Error` currently? Even if it was an AFError, you'd still have to cast it wouldn't you?

Reply:
This is an extension on `Result`, which is `Result<Success, Failure: Error>`, so we should use the generic here so the extension is useful on all `Result`s.

Reply:
Oh I follow now, good point. I guess I was assuming people would use the `af` namespace on non-AFResult types.

Reply:
We decided not to vend any APIs publicly so the point is moot.